# Statistics for Data Science — Week 3 Assignment
## Correlation, Least-Squares Regression, Residuals & Inference

**Objective:** Implement bivariate linear regression analysis in pure Python without using external libraries (such as NumPy, Pandas, or Scikit-learn).

---

### Dataset
Advertising spend $x$ (£000) against sales $y$ (£000) for six consecutive months:

| Month | Advertising Spend $x$ (£000) | Sales $y$ (£000) |
| :---: | :---: | :---: |
| 1 | 1 | 3 |
| 2 | 2 | 5 |
| 3 | 3 | 4 |
| 4 | 4 | 8 |
| 5 | 5 | 10 |
| 6 | 6 | 12 |

**Given:**
- Mean of $x$: $\bar{x} = 3.5$
- Mean of $y$: $\bar{y} = 7.0$


In [1]:
# Define dataset using standard Python lists (No external libraries)
x = [1, 2, 3, 4, 5, 6]
y = [3, 5, 4, 8, 10, 12]
n = len(x)

# Calculate sample means
x_bar = sum(x) / n
y_bar = sum(y) / n

print(f"Sample size (n): {n}")
print(f"Calculated x̄: {x_bar:.4f} (Given: 3.5)")
print(f"Calculated ȳ: {y_bar:.4f} (Given: 7.0)")


Sample size (n): 6
Calculated x̄: 3.5000 (Given: 3.5)
Calculated ȳ: 7.0000 (Given: 7.0)


## Part A — Correlation, the Line, Residuals & Interpretation

### Question 1: Calculate the Pearson Correlation Coefficient $r$

**Formulas:**
$$S_{xy} = \sum (x - \bar{x})(y - \bar{y})$$
$$S_{xx} = \sum (x - \bar{x})^2$$
$$S_{yy} = \sum (y - \bar{y})^2$$
$$r = \frac{S_{xy}}{\sqrt{S_{xx} \times S_{yy}}}$$


In [2]:
# Question 1: Pearson correlation coefficient r

# Calculate sum of products and sums of squared deviations
Sxy = sum((xi - x_bar) * (yi - y_bar) for xi, yi in zip(x, y))
Sxx = sum((xi - x_bar) ** 2 for xi in x)
Syy = sum((yi - y_bar) ** 2 for yi in y)

# Pearson correlation coefficient r (using pure Python exponentiation for square root)
r = Sxy / ((Sxx * Syy) ** 0.5)

print("--- Question 1 Calculations ---")
print(f"Sxy = Σ(x - x̄)(y - ȳ) = {Sxy:.4f}")
print(f"Sxx = Σ(x - x̄)²       = {Sxx:.4f}")
print(f"Syy = Σ(y - ȳ)²       = {Syy:.4f}")
print(f"r   = Sxy / √(Sxx*Syy) = {r:.6f}")
print(f"Rounded r             = {round(r, 4)}")


--- Question 1 Calculations ---
Sxy = Σ(x - x̄)(y - ȳ) = 32.0000
Sxx = Σ(x - x̄)²       = 17.5000
Syy = Σ(y - ȳ)²       = 64.0000
r   = Sxy / √(Sxx*Syy) = 0.956183
Rounded r             = 0.9562


### Question 2: Find the Least-Squares Regression Line

**Formulas:**
$$\text{Slope } b = \frac{S_{xy}}{S_{xx}}$$
$$\text{Intercept } a = \bar{y} - b\bar{x}$$
$$\text{Regression Equation: } \hat{y} = a + bx$$


In [3]:
# Question 2: Slope and Intercept of Regression Line

# Slope b
b = Sxy / Sxx

# Intercept a
a = y_bar - (b * x_bar)

print("--- Question 2 Calculations ---")
print(f"Slope (b)     = Sxy / Sxx = {Sxy:.1f} / {Sxx:.1f} = {b:.6f} ≈ {b:.3f}")
print(f"Intercept (a) = ȳ - b(x̄)  = {y_bar:.1f} - ({b:.4f} * {x_bar:.1f}) = {a:.6f} ≈ {a:.3f}")
print(f"Regression Line Equation: ŷ = {a:.4f} + {b:.4f}x")
print(f"In rounded form:          ŷ = {round(a, 3)} + {round(b, 3)}x")


--- Question 2 Calculations ---
Slope (b)     = Sxy / Sxx = 32.0 / 17.5 = 1.828571 ≈ 1.829
Intercept (a) = ȳ - b(x̄)  = 7.0 - (1.8286 * 3.5) = 0.600000 ≈ 0.600
Regression Line Equation: ŷ = 0.6000 + 1.8286x
In rounded form:          ŷ = 0.6 + 1.829x


### Question 3: Calculate $R^2$ and Interpretation

**Formula:**
$$R^2 = r^2$$


In [4]:
# Question 3: Coefficient of Determination R²
r_squared = r ** 2
r_squared_pct = r_squared * 100

print("--- Question 3 Calculations ---")
print(f"R² = r² = ({r:.6f})² = {r_squared:.6f}")
print(f"R² as percentage     = {r_squared_pct:.2f}% (≈ {round(r_squared_pct, 1)}%)")


--- Question 3 Calculations ---
R² = r² = (0.956183)² = 0.914286
R² as percentage     = 91.43% (≈ 91.4%)


**What does this percentage tell you about the variation in sales?**
- **Interpretation:** $R^2 = 91.43\%$ indicates that approximately **$91.4\%$ of the total variation in monthly sales ($y$) is explained by advertising expenditure ($x$)** via the linear regression model.
- Only the remaining **$8.57\%$ of the variation** is unexplained by the model (attributed to residual variance, random factors, or unobserved market conditions).


### Question 4: Calculate Adjusted $R^2$

**Formula:**
$$R^2_{\text{adj}} = 1 - (1 - R^2)\frac{n - 1}{n - k - 1}$$
where $n = 6$ (number of observations) and $k = 1$ (number of independent predictor variables).


In [5]:
# Question 4: Adjusted R²
k = 1  # single predictor: advertising spend

r_squared_adj = 1 - (1 - r_squared) * (n - 1) / (n - k - 1)
r_squared_adj_pct = r_squared_adj * 100

print("--- Question 4 Calculations ---")
print(f"n = {n}, k = {k}")
print(f"Degrees of freedom (total): n - 1 = {n - 1}")
print(f"Degrees of freedom (error): n - k - 1 = {n - k - 1}")
print(f"R²_adj = 1 - (1 - {r_squared:.6f}) * ({n - 1} / {n - k - 1})")
print(f"R²_adj = {r_squared_adj:.6f}")
print(f"R²_adj as percentage = {r_squared_adj_pct:.2f}% (≈ {round(r_squared_adj_pct, 1)}%)")


--- Question 4 Calculations ---
n = 6, k = 1
Degrees of freedom (total): n - 1 = 5
Degrees of freedom (error): n - k - 1 = 4
R²_adj = 1 - (1 - 0.914286) * (5 / 4)
R²_adj = 0.892857
R²_adj as percentage = 89.29% (≈ 89.3%)


**Why is Adjusted $R^2$ slightly different from ordinary $R^2$?**
- **Ordinary $R^2$** measures sample variance explained and naturally tends to overestimate the explanatory power in small samples ($n = 6$). It never decreases when additional predictors are added, even if they add no real explanatory value.
- **Adjusted $R^2$** introduces a penalty for the number of explanatory variables ($k$) relative to the sample size by weighting with degrees of freedom $\left(\frac{n-1}{n-k-1}\right)$. This yields an unbiased estimate of the population variance explained.
- Here, because $n$ is small ($n = 6$), the adjustment $\frac{5}{4} = 1.25$ slightly reduces the value from $91.43\%$ to $89.29\%$.


### Question 5: Calculate Predicted Values and Residuals

**Formula:**
$$\hat{y} = a + bx$$
$$\text{Residual } e = y - \hat{y}$$
Check that: $\sum e \approx 0$


In [6]:
# Question 5: Predicted Values ŷ, Residuals e, and Summary Table

y_hat_list = []
residual_list = []

header = f"{'Month':<6} {'x (£000)':<10} {'Actual y':<10} {'Predicted ŷ':<14} {'Residual e = y - ŷ':<18}"
print(header)
print("-" * len(header))

for i in range(n):
    xi = x[i]
    yi = y[i]
    y_hat_i = a + b * xi
    e_i = yi - y_hat_i
    
    y_hat_list.append(y_hat_i)
    residual_list.append(e_i)
    
    print(f"{i+1:<6} {xi:<10} {yi:<10.1f} {y_hat_i:<14.4f} {e_i:<18.4f}")

sum_x = sum(x)
sum_y = sum(y)
sum_y_hat = sum(y_hat_list)
sum_e = sum(residual_list)

print("-" * len(header))
print(f"{'Σ':<6} {sum_x:<10} {sum_y:<10.1f} {sum_y_hat:<14.4f} {sum_e:<18.10f}")
print("-" * len(header))

print(f"\nVerification check: Σe = {sum_e:.10f} ≈ 0 (Holds exactly true!)")


Month  x (£000)   Actual y   Predicted ŷ    Residual e = y - ŷ
--------------------------------------------------------------
1      1          3.0        2.4286         0.5714            
2      2          5.0        4.2571         0.7429            
3      3          4.0        6.0857         -2.0857           
4      4          8.0        7.9143         0.0857            
5      5          10.0       9.7429         0.2571            
6      6          12.0       11.5714        0.4286            
--------------------------------------------------------------
Σ      21         42.0       42.0000        -0.0000000000     
--------------------------------------------------------------

Verification check: Σe = -0.0000000000 ≈ 0 (Holds exactly true!)


---
## Part B — Inference & Logical Understanding

### Question 6: Interpret the Correlation
**Question:** The value of $r$ is positive and close to 1. What does this tell you about the relationship between advertising spend and sales?

**Answer:**
- The Pearson correlation coefficient is **$r \approx +0.956$**.
- Because $r > 0$ and very close to $+1$, it demonstrates an **extremely strong, positive linear relationship** between monthly advertising expenditure and sales.
- In practical terms, this means that as advertising spend increases across the six months, sales increase in a highly consistent, proportional linear pattern.


### Question 7: Interpret the Slope
**Question:** Suppose your regression line gives $b = 1.829$. Explain what this means in the context of advertising and sales. Remember that both $x$ and y are measured in £000.

**Answer:**
- The slope is $b = 1.829$ (or $b \approx 1.8286$).
- Since $x$ and $y$ are measured in thousands of pounds (£000), a unit increase of $\Delta x = 1$ represents an additional **£1,000** in advertising spend.
- **Interpretation:** For every additional **£1,000** spent on advertising, monthly sales are estimated to increase on average by **£1,829** (or $1.829 \times \text{£}1,000$).


### Question 8: Interpret $R^2$
**Question:** Suppose you obtain $R^2 \approx 0.914$. Explain this result in simple business language. Do not simply say “91.4% is explained.” Explain what is being explained.

**Answer:**
- **In simple business language:** Sales vary month-to-month across the six-month observation period (ranging from £3,000 to £12,000). About **$91.4\%$ of these differences and fluctuations in monthly sales** can be directly accounted for by how much money was allocated to advertising in those months.
- The remaining **$8.6\%$ of sales fluctuation** cannot be explained by advertising expenditure alone; it is caused by other unmeasured business factors (such as competitor campaigns, word-of-mouth, seasonality, product availability) or random variation.


### Question 9: Compare $R^2$ and Adjusted $R^2$
**Question:** Your $R^2$ and Adjusted $R^2$ are slightly different ($R^2 = 91.43\%$, $R^2_{\text{adj}} = 89.29\%$). What does this difference tell you about the usefulness of the model with the current number of predictors?

**Answer:**
- The difference between $R^2$ ($91.43\%$) and $R^2_{\text{adj}}$ ($89.29\%$) is very small (only **$2.14\%$**).
- This minor drop indicates that advertising expenditure is a **genuinely useful and powerful predictor**. 
- Even after penalizing for the small sample size ($n=6$) and loss of degrees of freedom, the model still explains nearly $90\%$ of the variance. It confirms that the high $R^2$ is not an artifact of model over-parameterization or meaningless predictors.


### Question 10: Positive Residual or Negative Residual?
**Question:** For one month, suppose $y = 4$ and $\hat{y} = 6.086$. Calculate the residual and explain what it means. Is actual sales above or below the predicted value?


In [7]:
# Question 10: Residual Calculation for Month 3
y_actual_q10 = 4.0
y_pred_q10 = 6.086
residual_q10 = y_actual_q10 - y_pred_q10

print("--- Question 10 Calculations ---")
print(f"Actual sales (y)     = £{y_actual_q10:.3f} thousand (£{int(y_actual_q10*1000):,})")
print(f"Predicted sales (ŷ)  = £{y_pred_q10:.3f} thousand (£{int(y_pred_q10*1000):,})")
print(f"Residual (e = y - ŷ) = {residual_q10:.3f} thousand (£{int(residual_q10*1000):,})")


--- Question 10 Calculations ---
Actual sales (y)     = £4.000 thousand (£4,000)
Predicted sales (ŷ)  = £6.086 thousand (£6,086)
Residual (e = y - ŷ) = -2.086 thousand (£-2,086)


**Answer:**
- **Residual value:** $e = 4 - 6.086 = -2.086$ (£000).
- This is a **negative residual**.
- **Meaning:** Actual sales (£4,000) were **below** the predicted value (£6,086) by **£2,086**. In other words, the regression model overpredicted sales for that specific month.


### Question 11: Business Inference
**Question:** A manager says: “If we increase advertising, sales will definitely increase.” Do you agree with this statement? Explain your reasoning using correlation, regression, and the meaning of causation.

**Answer:**
**I do not agree with the manager’s definitive statement.**

**Reasoning:**
1. **Correlation and Regression Do Not Prove Causation:** High correlation ($r = 0.956$) and a fitted regression line establish a strong statistical association, but they do not prove that advertising directly causes the increase in sales. A lurking third variable—such as seasonal demand or macroeconomic conditions—could be driving both advertising budget expansions and higher sales simultaneously.
2. **Deterministic vs. Probabilistic Nature:** The word *"definitely"* implies a guaranteed, deterministic outcome. Regression models provide only an *expected average trend* subject to random noise and unexplained variance ($8.6\%$ of variance remains unexplained).
3. **Diminishing Returns & Market Saturation:** In business, increasing ad spend beyond a certain point typically experiences diminishing marginal returns. Sales will not continue to grow indefinitely at £1,829 per £1,000 spent.


### Question 12: Prediction ($x = 7$)
**Question:** Using your regression equation, estimate sales when advertising expenditure is $x = 7$. Is this prediction interpolation or extrapolation? Explain why this distinction matters.


In [8]:
# Question 12: Prediction for x = 7
x_new = 7
y_pred_new = a + b * x_new

print("--- Question 12 Calculations ---")
print(f"Regression Equation: ŷ = {a:.4f} + {b:.4f} * x")
print(f"At x = {x_new}: ŷ = {a:.4f} + {b:.4f} * ({x_new}) = {y_pred_new:.4f} thousand (£{y_pred_new*1000:,.2f})")
print(f"Using rounded parameters (ŷ = 0.6 + 1.829x): ŷ = {0.6 + 1.829 * 7:.3f} thousand")


--- Question 12 Calculations ---
Regression Equation: ŷ = 0.6000 + 1.8286 * x
At x = 7: ŷ = 0.6000 + 1.8286 * (7) = 13.4000 thousand (£13,400.00)
Using rounded parameters (ŷ = 0.6 + 1.829x): ŷ = 13.403 thousand


**Answer:**
- **Estimated sales for $x = 7$:** $\hat{y} = 0.600 + 1.8286(7) = 13.400$ (£000), which equals **£13,400**.
- **Interpolation vs. Extrapolation:** This prediction is **extrapolation**.
- **Why this distinction matters:**
  - The observed advertising spend in the dataset ranges strictly from $x = 1$ to $x = 6$. The value $x = 7$ lies outside the observed domain.
  - *Interpolation* predicts within the known range where model linearity has been empirically observed.
  - *Extrapolation* assumes the exact same linear pattern holds true beyond observed limits. This is risky in business because market saturation, competitor counter-reactions, or supply bottlenecks can alter or flatten the relationship at higher expenditure levels.


### Question 13: Residual Reasoning
**Question:** Which month has the largest absolute residual? What does a large residual indicate about that month’s actual sales compared with the model’s prediction?


In [9]:
# Question 13: Find the largest absolute residual
abs_residuals = [abs(e) for e in residual_list]
max_abs_e = max(abs_residuals)
max_idx = abs_residuals.index(max_abs_e)

print("--- Question 13 Calculations ---")
for i, (xi, yi, yh, e) in enumerate(zip(x, y, y_hat_list, residual_list)):
    print(f"Month {i+1}: x = {xi}, Actual y = {yi:2.0f}, Predicted ŷ = {yh:6.4f}, Residual e = {e:7.4f}, |e| = {abs(e):6.4f}")

print(f"\nLargest absolute residual: |e| = {max_abs_e:.4f} at Month {max_idx + 1} (x = {x[max_idx]}, y = {y[max_idx]})")


--- Question 13 Calculations ---
Month 1: x = 1, Actual y =  3, Predicted ŷ = 2.4286, Residual e =  0.5714, |e| = 0.5714
Month 2: x = 2, Actual y =  5, Predicted ŷ = 4.2571, Residual e =  0.7429, |e| = 0.7429
Month 3: x = 3, Actual y =  4, Predicted ŷ = 6.0857, Residual e = -2.0857, |e| = 2.0857
Month 4: x = 4, Actual y =  8, Predicted ŷ = 7.9143, Residual e =  0.0857, |e| = 0.0857
Month 5: x = 5, Actual y = 10, Predicted ŷ = 9.7429, Residual e =  0.2571, |e| = 0.2571
Month 6: x = 6, Actual y = 12, Predicted ŷ = 11.5714, Residual e =  0.4286, |e| = 0.4286

Largest absolute residual: |e| = 2.0857 at Month 3 (x = 3, y = 4)


**Answer:**
- **Month 3** ($x = 3, y = 4$) has the largest absolute residual: $|e| = |-2.0857| \approx 2.086$ (£2,086).
- **What it indicates:** A large residual indicates that Month 3 represents the largest departure/discrepancy from the linear model. Actual sales (£4,000) substantially underperformed relative to the expected model prediction (£6,086).
- This suggests that unique external negative factors were at play during Month 3 (e.g., store stockouts, severe weather, negative press, or an aggressive promotional campaign launched by a competitor).


### Question 14: Logical Understanding
**Question:** An analyst says: “Because $R^2 = 91.4\%$, the model is $91.4\%$ accurate.” Is this a correct interpretation of $R^2$? Explain what is wrong with the statement.

**Answer:**
**No, this is an incorrect interpretation of $R^2$.**

**Explanation:**
1. **$R^2$ measures explained variance, not accuracy:** $R^2$ is the coefficient of determination, which represents the proportion of total variation in sales that is explained by the regression line relative to a baseline naive model (the mean $\bar{y}$).
2. **Accuracy has a different statistical definition:** Model accuracy typically refers to how close individual predictions are to actual values (e.g., Mean Absolute Percentage Error (MAPE), classification accuracy, or precision).
3. **High $R^2$ does not mean negligible errors:** A regression model can possess a very high $R^2$ (e.g., $91.4\%$) while still making substantial dollar errors on individual forecasts (e.g., Month 3 was off by £2,086 on an actual value of £4,000—a $52\%$ relative error!).


### Question 15: Final Conclusion
**Question:** Write a 2–3 sentence conclusion for the manager. Mention: (1) direction and strength of the relationship, (2) what the regression model suggests, (3) what $R^2$ tells us, and (4) one limitation of making a causal conclusion from these six observations.

**Answer:**
> **Executive Summary for Management:**  
> Our analysis reveals a **very strong positive linear relationship ($r = 0.956$)** between advertising spend and sales, with the regression model estimating that **each additional £1,000 spent on advertising generates an average sales increase of approximately £1,829** ($\hat{y} = 0.600 + 1.829x$). Furthermore, **advertising spend explains $91.4\%$ of the total monthly sales variation ($R^2 = 91.4\%$)**, indicating strong predictive alignment. However, because this analysis is based on **only six observations**, statistical association does not establish direct causation, and management should not assume higher advertising will guarantee sales growth without verifying confounding market factors and potential saturation limits.
